### DISCUSSION TITLE
Nextjs 14 app router webpack error

### DISCUSSION THREAD

**Discussion by pedrosousa13** [pmndrs/jotai#2453] [Bug report]
## Summary
When loading a simple app router app with next 14, there's a webpack error.

I simply am trying to use `useAtomValue` and get `Error: (0 , jotai_react__WEBPACK_IMPORTED_MODULE_2__.useAtomValue) is not a function`

## Link to reproduction
[CodeSandbox](https://codesandbox.io/p/devbox/nostalgic-wave-7q63nx?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cltpn13wb00072e660i62tmyr%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cltpn13wb00022e6680owl8rq%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cltpn13wb00042e6653zblkg0%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cltpn13wb00062e66eotnzh7u%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cltpn13wb00022e6680owl8rq%2522%253A%257B%2522id%2522%253A%2522cltpn13wb00022e6680owl8rq%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522cltpn13wb00062e66eotnzh7u%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cltpn13wb00052e66ttrutx3t%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A3000%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522cltpn13wb00062e66eotnzh7u%2522%252C%2522activeTabId%2522%253A%2522cltpn13wb00052e66ttrutx3t%2522%257D%252C%2522cltpn13wb00042e6653zblkg0%2522%253A%257B%2522id%2522%253A%2522cltpn13wb00042e6653zblkg0%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cltpn13wb00032e661e88cdbq%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%255D%252C%2522activeTabId%2522%253A%2522cltpn13wb00032e661e88cdbq%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

## Check List

Please do not ask questions in issues.

- [ ] I've already opened a [discussion](https://github.com/pmndrs/jotai/discussions) before opening this issue, or already discussed in other media.

Please include a minimal reproduction.

- [x] I've added a link to a [typescript playground](https://www.typescriptlang.org/play) or [codesandbox](https://codesandbox.io) with a minimal reproduction.

**Comment by dai-shi** [comment]
Feels like we get this report several times from nextjs users, but not all nextjs users.
Does anyone know how to fix?

**Reply by pedrosousa13** [reply]
I'll try to see if it can pinpoint the problem too. Just ran into it this morning.

**Reply by pedrosousa13** [reply]
I was checking and it only breaks when I'm using a server component. Maybe the rules for app router need to be better defined on the docs? mention that these can only be used in client components?

**Reply by dai-shi** [reply]
I see. Well, we have specified `"use client"`, so it should error more nicely.
I think it's a Nextjs-only bundling issue. I wonder if the issue is specific to Jotai.

**Reply by jet2jet** [reply]
This problem is not specific to Jotai; for example, when adding `'use client'` to a custom hook file and using it from a server component, a similar error will occur.

By the way, it may be nice that Jotai will depend to [`client-only` package](https://www.npmjs.com/package/client-only), indicating that Jotai cannot be used from server components.

**Reply by pedrosousa13** [reply]
that's a very good point, if it tells you the error outright then I can make a PR for it later.

**Reply by dai-shi** [reply]
I know how `client-only` works, but not a big fan of it. I would rather have "react-server" condition exports for libs. (but, it also needs some considerations, not any time soon.)

Still, I wonder why Next.js shows such an error. The error should be something like `useState` is not available.

**Reply by RosculescuCiprian** [reply]
I know this might be an old issue but I am using turbopack, the latest and greatest bundler from nextjs and I think it gives more of a clear error message
```Error: Attempted to call useHydrateAtoms() from the server but useHydrateAtoms is on the client. It's not possible to invoke a client function from the server, it can only be rendered as a Component or passed to props of a Client Component.```

**Reply by val-samonte** [reply]
I also tried using `useHydrateAtoms` (jotai 2.8.0, nextjs 14.2.3) and I am hitting this webpack issue. 
![image](https://github.com/pmndrs/jotai/assets/767060/bc7edf96-6ea4-48a5-871a-4ed6608fa05a)

So not sure how can I use `useHydrateAtoms` now since it needs to be on the SSR side of things.

**Reply by val-samonte** [reply]
Found the possible offender. I did copy Provider and useHydrateAtoms to my source, and found out Nextjs is complaining about the `useRef` and `createContext`.

![image](https://github.com/pmndrs/jotai/assets/767060/e21d29d1-2bff-44d2-bb4e-311e3c3542e6)

**Reply by dai-shi** [reply]
Can you try v2.5.1 if it shows a different error message?
In any case, you can't use Jotai's any use* hooks in server components because there are no states on the server. SSR is fine.
